### DISCUSSION TITLE
signing secret hash

### DISCUSSION THREAD

**Discussion by HEAVYPOLY** [pocketbase/pocketbase#3273] [Q&A]
lemonsqueezy webhooks are signed with a signing secret like this:  https://docs.lemonsqueezy.com/help/webhooks#signing-requests

On the pocketbase side I'm trying to read the data like this:
Is there an alternative to crypto package (which is only nodejs, doesn't seem to work)
```javascript
const crypto = require('crypto');
function verifySignature(request, secret) {
  const receivedSignature = request.headers['x-signature'];

  if (!receivedSignature) {
    // If no signature is provided, reject the request
    return false;
  }

  const hmac = crypto.createHmac('sha256', secret);
  const expectedSignature = hmac.update(JSON.stringify(request.body)).digest('hex');

  return crypto.timingSafeEqual(Buffer.from(receivedSignature, 'hex'), Buffer.from(expectedSignature, 'hex'));
}

routerAdd("POST", "/asdf", (c) => {
  // Verify the signature before processing the request
  if (!verifySignature(c.request, SIGNING_SECRET)) {
    // If the signature doesn't match, reject the request
    return c.json(403, { "message": "Unauthorized" });
  }

  const data = $apis.requestInfo(c).data;
  console.log(data);
  return c.json(200, { "message": data });
});
```

**Comment by ganigeorgiev** [comment]
There is no HMAC crypto primitive currently exposed in the JSVM bindings. I'll consider adding it sometime later this week as part of v0.18.2 since I also forgot to document the recent `$security.sha256()`, `$security.sha512()` and `$security.md5()` methods.

As a side-note - keep in mind that the JS `pb_hooks` implementation has a limitation that prevent us to use functions and variables declared outside of the handler scope (see https://pocketbase.io/docs/js-overview/#handlers-scope).

**Reply by HEAVYPOLY** [reply]
Ok, sounds good!  Do you recommend using go instead for now?  Is HMAC crypto already available on the go side?

**Reply by ganigeorgiev** [reply]
[`crypto/hmac`](https://pkg.go.dev/crypto/hmac) is part of the standard Go library.
Whether you should use Go - it depends whether you are familiar with the language and want to switch (see also the recommendations in https://pocketbase.io/docs/use-as-framework/), but in general you'll have more control over the application flow when using Go. The JS `pb_hooks` are technically wrapped Go code under the hood and execute more or less the same code (_as long as a binding exists_).

**Reply by HEAVYPOLY** [reply]
I've been wanting to learn Go, maybe this is a good excuse.  Thanks again.

**Reply by HEAVYPOLY** [reply]
Followed docs to run custom pb go server locally which worked perfectly.  Now I'm wondering how to deploy to production (pockethost).  Does main.go go into pb_hooks?

**Reply by ganigeorgiev** [reply]
When using Go you need to build your program with `go build`. This will create an executable that you can upload to your server and start it as usual.

**But I don't think PocketHost allows users to upload and start custom binaries**, aka. you need your own VPS or deploy to a service that allows running custom executables (_fly, hop, etc.; most of those services require building a custom Docker image that includes your binary; there are multiple ways to do it but you'll have to research it on your own if you want to go through this route_).

**Reply by HEAVYPOLY** [reply]
Ok, in this case I'll wait for HMAC on the js side. Thank you!  Also on the subject of security is it safe to include secret key strings directly inside main.pb.js ?

**Reply by ganigeorgiev** [reply]
With [PocketBase v0.18.2](https://github.com/pocketbase/pocketbase/releases/tag/v0.18.2), the following crypto primitives were added:

```js
// HMAC with SHA256
$security.hs256("hello", "secret")

// HMAC with SHA512
$security.hs512("hello", "secret")

// compare 2 strings with a constant time
$security.equal(hash1, hash2)
```
I haven't tested it with Lemon Squeezy but based on their documentation I think the above should suffice.


>  is it safe to include secret key strings directly inside main.pb.js ?

Yes. Optionally you can also move the secret in a `pb_hooks/config.json` file and load it from inside the handler in case you want to share the same secret with multiple hooks:
```js
routerAdd("POST", "/example", (c) => {
  const config = require(`${__hooks}/config.json`)
  
  ...
})
```

**Reply by HEAVYPOLY** [reply]
Amazing, thank you!  I will test and report back.

**Comment by HEAVYPOLY** [comment]
Getting invalid signature, does this look correct?
```javascript
const request = $apis.requestInfo(c);
const data = request.data        
const signature = request.headers["X-Signature"]
const computedSignature = $security.hs256(JSON.stringify(data), SIGNING_SECRET)
const isSignatureValid = $security.equal(computedSignature, signature)

if (!isSignatureValid) {
	return c.json(403, { "error": "Invalid signature" })
}
```

**Reply by ganigeorgiev** [reply]
The headers name in the `requestInfo` result is always normalized (see https://pocketbase.io/docs/js-routing/#reading-request-headers), aka. try with `request.headers['x_signature']`.

If changing the header name does't work, then I guess another thing that also could affect it is that Go maps doesn't preserve the insertion order of the fields and in your case this could matter because the final signature would differ. The format of the serialized json also could affect it, aka. `JSON.stringify(data)` will return "normalized" json string and will not preserve spaces and new lines between the keys and values. Or in other words you may need the **raw body** but we currently don't have an easy way to retrieve it from within the JS hooks (_I'll think a little more on it sometime later today_).

**Reply by HEAVYPOLY** [reply]
Tried ```const signature = request.headers["x_signature"]``` but no luck.  The lemonsqueezy data looks like this:
<img width="229" alt="image" src="https://github.com/pocketbase/pocketbase/assets/24432677/e587d1da-464a-47e9-861a-056c58445b35">
<img width="202" alt="image" src="https://github.com/pocketbase/pocketbase/assets/24432677/94b3f811-e9c8-47e9-9460-346d46fcbdf8">

**Reply by ganigeorgiev** [reply]
I'll submit sometime later today another minor release to allow reading the raw body.

**Reply by ganigeorgiev** [reply]
@HEAVYPOLY Could you try with [PocketBase v0.18.3](https://github.com/pocketbase/pocketbase/releases/tag/v0.18.3) and the new [`readerToString()`](https://pocketbase.io/jsvm/functions/readerToString.html) helper?

Your action could look something like this:

```js
routerAdd("POST", "/example", (c) => {
    const secret = "..."

    // read request data
    const info = $apis.requestInfo(c)
    const signature = info.headers["x_signature"] || '';
    const rawBody = readerToString(c.request().body) // note: after this call the body can't be read anymore

    // generate HMAC with SHA256 hash
    const hash = $security.hs256(rawBody, secret)

    // validate the hash against the header signature
    const isValid = $security.equal(hash, signature)
    if (!isValid) {
        throw new BadRequestError("Invalid webhook signature.")
    }

    // do something with the hook payload
    console.log(info.data);
})
```

If the above still doesn't work, double check that all variables has non-empty values.

**Reply by HEAVYPOLY** [reply]
SUCCESS! thank you!

**Reply by danbritt** [reply]
Thank you for adding this new helper! I also had a need to access the raw body of the request and discovered this thread while trying to figure out how.
